infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java

```diff
@@ -140,17 +147,32 @@ public KubernetesInternalRuntime(
     this.executor = sharedPool.getExecutor();
     this.runtimeStates = runtimeStates;
     this.machines = machines;
+    this.toolingProvisioner = toolingProvisioner;
+    this.kubernetesEnvironmentProvisioner = kubernetesEnvironmentProvisioner;
+    this.internalEnvironmentProvisioners = internalEnvironmentProvisioners;
     this.startSynchronizer = startSynchronizerFactory.create(context.getIdentity());
   }
 
   @Override
   protected void internalStart(Map<String, String> startOptions) throws InfrastructureException {
-    KubernetesRuntimeContext<? extends KubernetesEnvironment> context = getContext();
+    KubernetesRuntimeContext<E> context = getContext();
     String workspaceId = context.getIdentity().getWorkspaceId();
     try {
       startSynchronizer.setStartThread();
       startSynchronizer.start();
 
+      toolingProvisioner.provision(context.getIdentity(), context.getEnvironment());
+
+      startSynchronizer.checkFailure();
```

Comment:
Since we have 3 different provisioners it is not clear what is the difference and what is the order. Eventually, we would be able to move 2 of them to the infra abstract code, but, for now, it would be great to have a comment in the code that would explain:
- which provisioner is specific to what
- provisioners order and the reason

Reply:
Added